tests/cloudpickle_test.py

```diff
@@ -2321,6 +2325,47 @@ def __type__(self):
         o = MyClass()
         pickle_depickle(o, protocol=self.protocol)
 
+    def test_pickle_module_registered_for_pickling_by_value(self):
+        try:
+            reference = cloudpickle.dumps(an_external_function, protocol=cloudpickle.DEFAULT_PROTOCOL)
+            f1 = cloudpickle.loads(reference)
+
+            register_pickle_by_value("tests.external")
+            deep = cloudpickle.dumps(an_external_function, protocol=cloudpickle.DEFAULT_PROTOCOL)
+            f2 = cloudpickle.loads(deep)
+            unregister_pickle_by_value("tests.external")
+
+            # Ensure the serialisation is not the same
+            assert reference != deep
+            assert len(deep) > len(reference)
+            assert b"return_a_string" in deep
+            assert b"return_a_string" not in reference
+            assert f1() == f2()
+        finally:
+            _PICKLE_BY_VALUE_MODULES.clear()
+
+    def test_pickle_entire_module_by_value(self):
+        import _cloudpickle_testpkg.mod as m
+
+        try:
+            reference = cloudpickle.dumps(m, protocol=cloudpickle.DEFAULT_PROTOCOL)
+            m1 = cloudpickle.loads(reference)
+
+            register_pickle_by_value("_cloudpickle_testpkg")
+            deep = cloudpickle.dumps(m, protocol=cloudpickle.DEFAULT_PROTOCOL)
+            m2 = cloudpickle.loads(deep)
+            unregister_pickle_by_value("_cloudpickle_testpkg")
+        
+            # Ensure the serialisation is not the same
+            assert reference != deep
+            assert len(deep) > len(reference)
+            assert b"hello from a package" in deep
+            assert b"hello from a package" not in reference
+            assert m1.package_function() == m2.package_function()
```

Comment:
I think you are currently trying to pickle the **module** `_cloudpickle_testpkg.mod` and not the **package** `_cloudpickle_testpkg` - this is why the three assertions are currently failing.  You can look for objects from `_cloudpickle_testpkg.mod` instead :)

Reply:
Yeah I was trying to see if I pickled the submodule (without the typing) if it would work. If I make it back to `import _cloudpickle_testpkg as m` I get the same TypeError and two ImportErrors:

```
FAILED tests/cloudpickle_test.py::CloudPickleTest::test_relative_import_inside_function - ImportError: __import__ not found
FAILED tests/cloudpickle_test.py::Protocol2CloudPickleTest::test_relative_import_inside_function - ImportError: __import__ not found
FAILED tests/cloudpickle_test.py::test_register_pickle_by_value - TypeError: cannot check importability of TypeVar instances
```

Reply:
OK, the `__import__` error may be tricky to get. Feel free to put this on pause, I'll try to get this done today.

Reply:
Can mate. Apologies for needing some hand holding to get it done. Let me know if theres anything further I can do which will assist that wouldn't just be more efficient for you to jump in and do yourself haha. Appreciate the help so far :)

Reply:
The import error will be fixed by #426.

Reply:
Is there a way I can pull in #426 into my changes? Still have to figure out the TypeError part somehow and then hopefully its all good

Reply:
You can try to rebase this branch (or to be safer, a copy of it at first) on #426.

Reply:
> Apologies for needing some hand holding to get it done.

No worries, thank you for being responsive all the way through.

Reply:
I'm going to wait <24 hours for feedback on #426, and otherwise will merge so that we can finish up the work of this PR ASAP.

Reply:
That was a nasty bug...